### ISSUE TITLE
[native-image] Segfault during DMA operation

### ISSUE THREAD

**Issue by Jotschi** [oracle/graal#529]
I'm seeing the listed segmentation fault during a DNA operation on a Netty PooledUnsafeDirectByteBuf instance. I'm not sure what actually causes the issue. I tried to create a smaller reproducer to see whether the way DNA is handled by Netty is not compatible with substratevm (see DMATest.java).  That test is working just fine.

Maybe someone could give me a hint on how to proceed here. ~~I'm not that familiar with the Netty codebase but I think the lack of DNA support in native-images that I previously build could explain why the JVM is still outperforming native images.~~

## Reproducer

```
# Checkout
git clone git@github.com:Jotschi/vertx-graalvm-native-image-test.git
git checkout 95b310836cddfa5c2162ae829147f0d9e89a58ea

# Build app
mvn clean package
native-image \
 --verbose \
 --no-server \
 -H:Name=hello-world \
 -H:+ReportUnsupportedElementsAtRuntime \
 -Dfile.encoding=UTF-8 \
 -jar target/vertx-graalvm-native-image-test-0.0.1-SNAPSHOT.jar

# Run image
./vertx-graalvm-native-image-test-0.0.1-SNAPSHOT & sleep 2 ; curl http://localhost:8080/hello
```

## Trace

```
18:52:00.915 [main] INFO  de.jotschi.examples.Runner - Starting server for: http://localhost:8080/hello
Using recycler: io.netty.buffer.PooledUnsafeDirectByteBuf$1
io.netty.buffer.PooledUnsafeDirectByteBuf
[ [ SubstrateSegfaultHandler caught signal 11 ] ]

General Purpose Register Set Values: 

  RAX 00000000ffffffff
  RBX 00007f0dfc24a090
  RCX 0000000000000000
  RDX 0000000000000001
  RBP 0000000000000000
  RSI 0000000000000000
  RDI 00007f0dfc24a090
  RSP 00007f0dfcbfea40
  R8  00007f0dfec01c60
  R9  0000561e9cd9bd50
  R10 00007f0dff726180
  R11 0000000000000000
  R12 0000000000000010
  R13 0000000000000003
  R14 000000000000000a
  R15 00007f0de80008c0
  EFL 0000000000010202
  RIP 0000561e9bfdfa9b
  

JavaFrameAnchor dump:

  No anchors
  
DeoptStubPointer address: 0000000000000000

TopFrame info:

  Lookup TotalFrameSize in CodeInfoTable:
  SourceTotalFrameSize 32
  
VMThreads info:

  VMThread 00007f0de80008c0  STATUS_IN_JAVA (safepoints disabled)
  VMThread 00007f0df40008c0  STATUS_IN_NATIVE
  VMThread 00007f0df00008c0  STATUS_IN_NATIVE
  VMThread 00007f0df80008c0  STATUS_IN_NATIVE
  VMThread 0000561e9eb4b010  STATUS_IN_NATIVE
  
VM Thread State for current thread 00007f0de80008c0:

  0 (8 bytes): com.oracle.svm.core.genscavenge.PinnedAllocatorImpl.openPinnedAllocator = null
  8 (8 bytes): com.oracle.svm.core.heap.NoAllocationVerifier.openVerifiers = null
  16 (8 bytes): com.oracle.svm.core.jdk.IdentityHashCodeSupport.hashCodeGeneratorTL = java.util.SplittableRandom  00007f0dfc201750
  24 (8 bytes): com.oracle.svm.core.snippets.SnippetRuntime.currentException = null
  32 (8 bytes): com.oracle.svm.core.thread.JavaThreads.currentThread = io.vertx.core.impl.VertxThread  00007f0dfcc03480
  40 (8 bytes): com.oracle.svm.core.thread.ThreadingSupportImpl.activeTimer = null
  48 (32 bytes): com.oracle.svm.core.genscavenge.ThreadLocalAllocation.pinnedTLAB = 
    00007f0de80008f0: 0000000000000000 0000000000000000
    00007f0de8000900: 0000000000000000 0000000000000000
    
  
  80 (32 bytes): com.oracle.svm.core.genscavenge.ThreadLocalAllocation.regularTLAB = 
    00007f0de8000910: 00007f0dfc200000 00007f0dfc300000
    00007f0de8000920: 00007f0dfc2504b0 0000000000000000
    
  
  112 (8 bytes): com.oracle.svm.core.genscavenge.ThreadLocalAllocation.freeList = 0  0000000000000000
  120 (8 bytes): com.oracle.svm.core.posix.thread.PosixVMThreads.IsolateTL = 94598687557484880  0150150150150150
  128 (8 bytes): com.oracle.svm.core.stack.JavaFrameAnchors.lastAnchor = 0  0000000000000000
  136 (8 bytes): com.oracle.svm.core.thread.VMThreads.nextTL = 139697904945344  00007f0df40008c0
  144 (4 bytes): com.oracle.svm.core.thread.Safepoint.safepointRequested = -14731  ffffc675
  148 (4 bytes): com.oracle.svm.core.thread.Safepoint.safepointRequestedValueBeforeSafepoint = 0  00000000
  152 (4 bytes): com.oracle.svm.core.thread.VMOperationControl.isLockOwner = 0  00000000
  156 (4 bytes): com.oracle.svm.core.thread.VMThreads$StatusSupport.safepointsDisabledTL = 1  00000001
  160 (4 bytes): com.oracle.svm.core.thread.VMThreads$StatusSupport.statusTL = 1  00000001
  
VMOperation dump:

  No VMOperation in progress
  

RuntimeCodeCache dump:

  == [Recent RuntimeCodeCache operations: ]
  
  == [RuntimeCodeCache: 0 methods]
  
Deoptimizer dump:

  == [Recent Deoptimizer Events: 
  ]
  
Dump Counters:

  
Raw Stacktrace:

  00007f0dfcbfea40: 0000561e9d021ba8 0000000000000001
  00007f0dfcbfea50: 00007f0dfc24a090 0000561e9bf98e2e
  00007f0dfcbfea60: 00007f0dfc2503a8 0000561e9bf31cb5
  00007f0dfcbfea70: 0000561e9cdc1828 000000009c79cc20
  00007f0dfcbfea80: 00007f0dfc24a090 0000561e9c26d74c
  00007f0dfcbfea90: 00007f0dfcc02920 0000561e9c640eb4
  00007f0dfcbfeaa0: 0000001800000000 0000561e9cdc1828
  00007f0dfcbfeab0: 00007f0dfc24a090 00007f0dfc202168
  00007f0dfcbfeac0: 00007f0dfc201930 0000561e9c01ce4f
  00007f0dfcbfead0: 00007f0dfcc02a48 0000561e9c01b55b
  00007f0dfcbfeae0: 00007f0dfc250368 00007f0dfc24a090
  00007f0dfcbfeaf0: 00007f0dfc202168 0000561e9c01cf9b
  00007f0dfcbfeb00: 0000000000000000 0000561e9bfe3de5
  00007f0dfcbfeb10: 00007f0dfc24a090 0000561e9cde1bd0
  00007f0dfcbfeb20: 00007f0dfec02dc0 00007f0dfc202168
  00007f0dfcbfeb30: 00007f0dfc24a090 0000561e9c01b812
  00007f0dfcbfeb40: 0000561e9cdc1828 000000009c5778e5
  00007f0dfcbfeb50: 00007f0dfc250368 00007f0dfc24a090
  00007f0dfcbfeb60: 00007f0dfcc02d28 0000561e9c035d07
  00007f0dfcbfeb70: 000000009ce6e290 00007f0dfec02dc0
  00007f0dfcbfeb80: 00007f0dfc24a090 00007f0dfcc02d28
  00007f0dfcbfeb90: 00007f0dfcc02d28 0000561e9c01ce4f
  00007f0dfcbfeba0: 00007f0dfcc02a48 0000561e9c01b55b
  00007f0dfcbfebb0: 0000040000000026 00007f0dfc24a090
  00007f0dfcbfebc0: 00007f0dfcc02d28 0000561e9c01cf9b
  00007f0dfcbfebd0: 0000561e9c79c730 0000561e9c04039e
  00007f0dfcbfebe0: 00007f0dfc24a090 0000561e9cde1bd0
  00007f0dfcbfebf0: 00007f0dfec02dc0 00007f0dfcc02d28
  00007f0dfcbfec00: 00007f0dfc24a090 0000561e9c0820c1
  00007f0dfcbfec10: 00007f0dfec070b0 0000561e9c63567b
  00007f0dfcbfec20: 00007f0dfec02dc0 00007f0dfc24a090
  00007f0dfcbfec30: 0000561e9c024da0 0000561e9cdd0570
  00007f0dfcbfec40: 00007f0dfc202320 0000561e9c79c730
  00007f0dfcbfec50: 00007f0dfcc02c40 00007f0dfcc02e60
  00007f0dfcbfec60: 00007f0dfcc02ba0 0000561e9c089d1b
  00007f0dfcbfec70: 00007f0dfc201658 731ec1d2fc2022b0
  00007f0dfcbfec80: 00000001fec07228 0000561e9c4d6108
  00007f0dfcbfec90: 00007f0dfec07208 00007f0dfcc02ba0
  00007f0dfcbfeca0: 00007f0dfcc02a48 00007f0dfc201658
  00007f0dfcbfecb0: 00007f0dfec02dc0 0000561e9c08a659
  00007f0dfcbfecc0: 00007f0dfec07228 000000004360b1ba
  00007f0dfcbfecd0: 00007f0dfcbfed28 00007f0dfee12996
  00007f0dfcbfece0: 00007f0dfc201658 00007f0dfcc02a48
  00007f0dfcbfecf0: 00007f0dfc2022f0 0000561e9c4d58c0
  00007f0dfcbfed00: 00007f0dfec07290 00007f0dfec07290
  00007f0dfcbfed10: 00007f0dfec02dc0 0000561e9c08a1cb
  00007f0dfcbfed20: 000000019c089b40 0000000000004f88
  00007f0dfcbfed30: 00007f0dfec02dc0 0000561e9c08b31d
  00007f0dfcbfed40: 0000000000000000 0000561e9c378555
  00007f0dfcbfed50: 00000000000000c8 0000000000000032
  00007f0dfcbfed60: 00007f0d00000032 0000000000000032
  00007f0dfcbfed70: 000012846ef958f5 8000000000000000
  00007f0dfcbfed80: 00007f0dfec02dc0 0000561e9c2241ab
  00007f0dfcbfed90: 0000561e9cd9c048 00007f0dfcc029f0
  00007f0dfcbfeda0: 00000001fc201440 0000561e9d3eb8d0
  00007f0dfcbfedb0: 00007f0dfc201460 00007f0dff7281a7
  00007f0dfcbfedc0: 00007f0dfc2010c0 00000000fc201460
  00007f0dfcbfedd0: 00007f0dfc201460 0000561e9bf35f5e
  00007f0dfcbfede0: 0000000000000010 0000000000000000
  00007f0dfcbfedf0: 0000561e9d3f8a48 0000561e9cdc8828
  00007f0dfcbfee00: 0000561e9ce7e528 0000561e9ce7e200
  00007f0dfcbfee10: 0000561e9ce30420 0000561e9ce7df08
  00007f0dfcbfee20: 00007f0dfcc03368 0000561e9c21e54f
  00007f0dfcbfee30: 0000000a00000000 00007f0dfc2010c0
  
Stacktrace Stage0:

  RSP 00007f0dfcbfea40 RIP 0000561e9bfdfa9b FrameSize 32
  RSP 00007f0dfcbfea60 RIP 0000561e9bf98e2e FrameSize 48
  RSP 00007f0dfcbfea90 RIP 0000561e9c26d74c FrameSize 64
  RSP 00007f0dfcbfead0 RIP 0000561e9c01ce4f FrameSize 48
  RSP 00007f0dfcbfeb00 RIP 0000561e9c01cf9b FrameSize 64
  RSP 00007f0dfcbfeb40 RIP 0000561e9c01b812 FrameSize 48
  RSP 00007f0dfcbfeb70 RIP 0000561e9c035d07 FrameSize 48
  RSP 00007f0dfcbfeba0 RIP 0000561e9c01ce4f FrameSize 48
  RSP 00007f0dfcbfebd0 RIP 0000561e9c01cf9b FrameSize 64
  RSP 00007f0dfcbfec10 RIP 0000561e9c0820c1 FrameSize 96
  RSP 00007f0dfcbfec70 RIP 0000561e9c089d1b FrameSize 80
  RSP 00007f0dfcbfecc0 RIP 0000561e9c08a659 FrameSize 96
  RSP 00007f0dfcbfed20 RIP 0000561e9c08a1cb FrameSize 32
  RSP 00007f0dfcbfed40 RIP 0000561e9c08b31d FrameSize 80
  RSP 00007f0dfcbfed90 RIP 0000561e9c2241ab FrameSize 160
  RSP 00007f0dfcbfee30 RIP 0000561e9c21e54f FrameSize 32
  RSP 00007f0dfcbfee50 RIP 0000561e9c37adff FrameSize 32
  RSP 00007f0dfcbfee70 RIP 0000561e9bf75893 FrameSize 96
  RSP 00007f0dfcbfeed0 RIP 0000561e9bf37044 FrameSize 144
  
Stacktrace Stage1:

  RSP 00007f0dfcbfea40 RIP 0000561e9bfdfa9b  com.oracle.svm.core.code.ImageCodeInfo@0x561e9cd9c960 name = image code
  RSP 00007f0dfcbfea60 RIP 0000561e9bf98e2e  com.oracle.svm.core.code.ImageCodeInfo@0x561e9cd9c960 name = image code
  RSP 00007f0dfcbfea90 RIP 0000561e9c26d74c  com.oracle.svm.core.code.ImageCodeInfo@0x561e9cd9c960 name = image code
  RSP 00007f0dfcbfead0 RIP 0000561e9c01ce4f  com.oracle.svm.core.code.ImageCodeInfo@0x561e9cd9c960 name = image code
  RSP 00007f0dfcbfeb00 RIP 0000561e9c01cf9b  com.oracle.svm.core.code.ImageCodeInfo@0x561e9cd9c960 name = image code
  RSP 00007f0dfcbfeb40 RIP 0000561e9c01b812  com.oracle.svm.core.code.ImageCodeInfo@0x561e9cd9c960 name = image code
  RSP 00007f0dfcbfeb70 RIP 0000561e9c035d07  com.oracle.svm.core.code.ImageCodeInfo@0x561e9cd9c960 name = image code
  RSP 00007f0dfcbfeba0 RIP 0000561e9c01ce4f  com.oracle.svm.core.code.ImageCodeInfo@0x561e9cd9c960 name = image code
  RSP 00007f0dfcbfebd0 RIP 0000561e9c01cf9b  com.oracle.svm.core.code.ImageCodeInfo@0x561e9cd9c960 name = image code
  RSP 00007f0dfcbfec10 RIP 0000561e9c0820c1  com.oracle.svm.core.code.ImageCodeInfo@0x561e9cd9c960 name = image code
  RSP 00007f0dfcbfec70 RIP 0000561e9c089d1b  com.oracle.svm.core.code.ImageCodeInfo@0x561e9cd9c960 name = image code
  RSP 00007f0dfcbfecc0 RIP 0000561e9c08a659  com.oracle.svm.core.code.ImageCodeInfo@0x561e9cd9c960 name = image code
  RSP 00007f0dfcbfed20 RIP 0000561e9c08a1cb  com.oracle.svm.core.code.ImageCodeInfo@0x561e9cd9c960 name = image code
  RSP 00007f0dfcbfed40 RIP 0000561e9c08b31d  com.oracle.svm.core.code.ImageCodeInfo@0x561e9cd9c960 name = image code
  RSP 00007f0dfcbfed90 RIP 0000561e9c2241ab  com.oracle.svm.core.code.ImageCodeInfo@0x561e9cd9c960 name = image code
  RSP 00007f0dfcbfee30 RIP 0000561e9c21e54f  com.oracle.svm.core.code.ImageCodeInfo@0x561e9cd9c960 name = image code
  RSP 00007f0dfcbfee50 RIP 0000561e9c37adff  com.oracle.svm.core.code.ImageCodeInfo@0x561e9cd9c960 name = image code
  RSP 00007f0dfcbfee70 RIP 0000561e9bf75893  com.oracle.svm.core.code.ImageCodeInfo@0x561e9cd9c960 name = image code
  RSP 00007f0dfcbfeed0 RIP 0000561e9bf37044  com.oracle.svm.core.code.ImageCodeInfo@0x561e9cd9c960 name = image code
  
Full Stacktrace:

  
  RSP 00007f0dfcbfea60 RIP 0000561e9bf98e2e  [image code] io.netty.buffer.AbstractByteBuf.getByte(AbstractByteBuf.java:341)
  RSP 00007f0dfcbfea90 RIP 0000561e9c26d74c  [image code] io.vertx.core.http.impl.Http1xOrH2CHandler.channelRead(Http1xOrH2CHandler.java:37)
  RSP 00007f0dfcbfead0 RIP 0000561e9c01ce4f  [image code] io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
  RSP 00007f0dfcbfeb00 RIP 0000561e9c01cf9b  [image code] io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
  RSP 00007f0dfcbfeb40 RIP 0000561e9c01b812  [image code] io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
  RSP 00007f0dfcbfeb70 RIP 0000561e9c035d07  [image code] io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1359)
  RSP 00007f0dfcbfeba0 RIP 0000561e9c01ce4f  [image code] io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
  RSP 00007f0dfcbfebd0 RIP 0000561e9c01cf9b  [image code] io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
  RSP 00007f0dfcbfec10 RIP 0000561e9c0820c1  [image code] io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:935)
  RSP 00007f0dfcbfec10 RIP 0000561e9c0820c1  [image code] io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:141)
  RSP 00007f0dfcbfec70 RIP 0000561e9c089d1b  [image code] io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:645)
  RSP 00007f0dfcbfecc0 RIP 0000561e9c08a659  [image code] io.netty.channel.nio.NioEventLoop.processSelectedKeysPlain(NioEventLoop.java:545)
  RSP 00007f0dfcbfed20 RIP 0000561e9c08a1cb  [image code] io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:499)
  RSP 00007f0dfcbfed40 RIP 0000561e9c08b31d  [image code] io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:459)
  RSP 00007f0dfcbfed90 RIP 0000561e9c2241ab  [image code] io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:886)
  RSP 00007f0dfcbfee30 RIP 0000561e9c21e54f  [image code] io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
  RSP 00007f0dfcbfee50 RIP 0000561e9c37adff  [image code] java.lang.Thread.run(Thread.java:748)
  RSP 00007f0dfcbfee70 RIP 0000561e9bf75893  [image code] com.oracle.svm.core.posix.thread.PosixJavaThreads.pthreadStartRoutine(PosixJavaThreads.java:238)
  RSP 00007f0dfcbfeed0 RIP 0000561e9bf37044  [image code] com.oracle.svm.core.code.CEntryPointCallStubs.com_002eoracle_002esvm_002ecore_002eposix_002ethread_002ePosixJavaThreads_002epthreadStartRoutine_0028com_002eoracle_002esvm_002ecore_002eposix_002ethread_002ePosixJavaThreads_0024ThreadStartData_0029(generated:0)
  
Use runtime option -R:-InstallSegfaultHandler if you don't want to use SubstrateSegfaultHandler.

Bye bye ...
```

**Comment by Jotschi** [comment]
I assume this change in [RC6](http://www.graalvm.org/docs/release-notes/#10-rc6) is related to the issue:

> Direct byte buffer are no longer allowed in the image heap: We added a new verification during image generation to ensure that no direct or mapped byte buffers (MappedByteBuffer instances) are in the image heap. Such buffers have either a pointer to C memory or reference a file descriptor, i.e., native resources that are available during image generation but no longer at image runtime. Using such a buffer would lead to a segfault at runtime. We discovered this issue because Netty has a few direct buffers that are created in class initializers. It is necessary to delay the initialization of such classes to runtime.

**Comment by cstancu** [comment]
@Jotschi I think this issue should be fixed using the latest GraalVM 19.3.0-dev. You need to update your dependencies; this will bring in the latest Netty which includes the required Native Image config by default. See also the resolution on the related issue: https://github.com/oracle/graal/issues/471#issuecomment-537236133. Please reopen if you still have problems.
